### ISSUE TITLE
Allow .dockercfg credentials to be provided via the environment

### ISSUE THREAD

**Issue by rafecolton** [docker/docker#6141]
This would be super handy for things that automate container builds / pushes / pulls.

**Comment by fermayo** [comment]
+1

**Comment by jujhars13** [comment]
I'm just using this workaround for the moment and defining the `${HOME}` env variable before I run docker pull/run
`HOME=/root docker pull buto/our-private-image`

**Comment by spaungam** [comment]
#dibs

**Comment by duglin** [comment]
@spaungam please see #11675  
If you're thinking about doing something with .dockercfg please hold off for a bit

**Comment by rafecolton** [comment]
@duglin it looks like #11675 is about overwriting the `.dockercfg` file's location. This issue is about allowing the actual values themselves be read in from the environment, obviating the need for a `.dockercfg` file in certain cases

**Comment by rafecolton** [comment]
To clarify, the interface might look something like this:

```bash
DOCKER_CFG_USERNAME="foo" DOCKER_CFG_PASSWORD="bar" DOCKER_CFG_EMAIL="baz@example.com" docker push something/somewhere
```

or maybe

```bash
DOCKER_CFG_JSON="{ ... json encoded .dockercfg ... }" docker push something/somewhere
```

(maybe both!)

**Comment by duglin** [comment]
ok it wasn’t clear to me exactly which direction people were headed.  Specifying the .dockercfg values from an env var doesn’t overlap with specifying a new location for the file. One thing to keep in mind is whether env vars or cmd line options are better - or both.  We don’t want to go too far and duplicate everything though - need to find the right balance.  I’m not keen on specifying JSON in an env var - feels a bit too low level and at that point I would just write it to a file and use the (hopefully coming soon feature) of --config <file>

thanks
-Doug

**Comment by vdemeester** [comment]
I think this one is closed by #12856 :wink:.

**Comment by rafecolton** [comment]
Hi @vdemeester - this is actually completely different from #12856.  

#12856 allows the user to specify a config file *location*.  That still requires that a config file exist and be read from disk.  This is a request to allow the *values* from the `.dockercfg` file to be provided in the environment, requiring that *no file be read from disk*.

**Comment by vdemeester** [comment]
@rafecolton oh I see, my bad :blush:

**Comment by rafecolton** [comment]
Also to address @duglin's comment about env vars vs command line flags - if an option must be provided on the command line and cannot be provided in the environment, that just means I have to wrap my command in a script that reads the environment variable and passes it as a command line option.  It's annoying but not unprecedented

**Comment by duglin** [comment]
Revisiting an old one...
Rather than encouraging people to encode the entire config file in an env var, I think it might be better to look at a more natural way for people to specify the config file properties they want to declare/override.  For example, for each option in the config file perhaps there should be a cmd line option equivalent?  So, for the creds perhaps we just need to explore whether we should add `--login` and `--password` type of flags to the `push` and `pull` cmds? 

And if so, I'd prefer to explore that via independent issues/PRs where we can consider each new flag in isolation.  With that, I think we should close this issue and then people can open new ones if needed/wanted.

what do people think?

**Comment by j0hnsmith** [comment]
I'm looking for a way to to not use a file at all, so flags would work, but could get labourious.

**Comment by duglin** [comment]
yea, but one of the things I like about flags is that you can target the exact config field you want to change.  Specifying the entire config means people may then need to worry about merging their desired change with the existing config file - and I think that's even messier.

**Comment by duglin** [comment]
ping @rafecolton

**Comment by thaJeztah** [comment]
Note that there's also an open pull-request to use an external credentials store https://github.com/docker/docker/pull/20107

**Comment by rafecolton** [comment]
@duglin RE flags vs env vars:

> Also to address @duglin's comment about env vars vs command line flags - if an option must be provided on the command line and cannot be provided in the environment, that just means I have to wrap my command in a script that reads the environment variable and passes it as a command line option. It's annoying but not unprecedented

External credential store sounds like a neat idea and would probably address this need

**Comment by thaJeztah** [comment]
I'm closing this one; the external credential feature (https://github.com/docker/docker/pull/20107) has been implemented in 1.11, and although there's currently not a credential store that uses environment variables, it should be possible to implement one. I don't think we'll implement reading from environment variables as a core functionality. I think the discussion around adding more flags to `docker pull` / `docker push` is a different one, so if there's still a need for that, please open a separate issue.

Closing, but feel free to comment after I closed
